gensim/models/fasttext.py

```diff
@@ -0,0 +1,363 @@
+                logger.warning("The rule, if given, is only used to prune vocabulary during build_vocab() and is not stored as part of the model. ")
+                logger.warning("Model initialized without sentences. trim_rule provided, if any, will be ignored.")
+
+    def initialize_ngram_vectors(self):
+        self.wv = FastTextKeyedVectors()
+
+    def build_vocab(self, sentences, keep_raw_vocab=False, trim_rule=None, progress_per=10000, update=False):
+        self.scan_vocab(sentences, progress_per=progress_per, trim_rule=trim_rule)  # initial survey
+        if update:
+            if not len(self.wv.vocab):
+                raise RuntimeError("You cannot do an online vocabulary-update of a model which has no prior vocabulary. "
+                    "First build the vocabulary of your model with a corpus "
+                    "before doing an online update.")
+            self.old_vocab_len = len(self.wv.vocab)
+            self.old_hash2index_len = len(self.wv.hash2index)
+        self.scale_vocab(keep_raw_vocab=keep_raw_vocab, trim_rule=trim_rule, update=update)  # trim by min_count & precalculate downsampling
+        self.finalize_vocab(update=update)  # build tables & arrays
+        self.init_ngrams(update=update)
+
+    def init_ngrams(self, update=False):
+        if not update:
+            self.wv.ngrams = {}
+            self.wv.syn0_all = empty((len(self.wv.vocab) + self.bucket, self.vector_size), dtype=REAL)
+            self.syn0_all_lockf = ones((len(self.wv.vocab) + self.bucket, self.vector_size), dtype=REAL)
+
+            all_ngrams = []
+            for w, v in self.wv.vocab.items():
+                self.wv.ngrams_word[w] = Ft_Wrapper.compute_ngrams(w, self.min_n, self.max_n)
+                all_ngrams += self.wv.ngrams_word[w]
+
+            all_ngrams = list(set(all_ngrams))
+            self.num_ngram_vectors = len(self.wv.vocab) + len(all_ngrams)
+            logger.info("Total number of ngrams is %d", len(all_ngrams))
+
+            self.wv.hash2index = {}
+            ngram_indices = range(len(self.wv.vocab))  # keeping the first `len(self.wv.vocab)` rows intact
+            for i, ngram in enumerate(all_ngrams):
+                ngram_hash = Ft_Wrapper.ft_hash(ngram)
+                if ngram_hash in self.wv.hash2index:
+                    self.wv.ngrams[ngram] = self.wv.hash2index[ngram_hash]
+                else:
+                    ngram_indices.append(len(self.wv.vocab) + ngram_hash % self.bucket)
+                    self.wv.hash2index[ngram_hash] = i + len(self.wv.vocab)
+                    self.wv.ngrams[ngram] = self.wv.hash2index[ngram_hash]
+
+            self.wv.syn0_all = self.wv.syn0_all.take(ngram_indices, axis=0)
+            self.reset_ngram_weights()
+        else:
+            new_vocab_len = len(self.wv.vocab)
+            for ngram, idx in self.wv.hash2index.items():
+                self.wv.hash2index[ngram] = idx + new_vocab_len - self.old_vocab_len
```

Comment:
I think it might be a good idea to have two separate matrices, one for storing the vectors for the `<word>` tokens, and one for the subwords. Along with renaming the variable to more intuitive names (we don't really need to follow the `syn0` `syn1` nomenclature here), that should make code much cleaner.

It also makes resizing much easier.

Reply:
Done